### PR TITLE
Suggest ballerina central packages when they're partially typed

### LINKED ISSUE #40816: [Improvement]: Suggest ballerina central packages when they're partially typed
### Description

Currently when a central package is partially typed, we don't suggest the relevant completions that start with what has been typed. we can fix this by improving the `importDeclarationNodeContext` 

### Steps to Reproduce

_No response_

### Affected Version(s)

_No response_

### OS, DB, other environment details and versions

_No response_

### Related area

-> Editor

### Related issue(s) (optional)

#40125

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by mindula**
## Purpose
This PR adds the implementation to suggest ballerina central packages when partially typed and changes the work done progress notification.

Fixes #40816, #41088

## Samples

Screenshot 2023-08-07 at 23 32 14

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/ImportDeclarationNodeContext.java [hunk 1]
```diff
@@ -318,30 +319,55 @@ private ArrayList<LSCompletionItem> moduleNameContextCompletions(BallerinaComple
...
-                completionItem.getCompletionItem().setAdditionalTextEdits(additionalEdits);
+        List<LSPackageLoader.ModuleInfo> moduleList;
+
+        if (orgName.equals("ballerinax")) {
+            List<String> packageList = new ArrayList<>();
+            String prefix = context.getTokenAtCursor().text();
+            moduleList = LSPackageLoader.getInstance(serverContext).getCentralPackages(serverContext);
+            moduleList.forEach(ballerinaPackage -> {
+                String packageName = ballerinaPackage.packageName().value();
+                packageList.add(packageName);
```
Comment (KavinduZoysa) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/ImportDeclarationNodeContext.java:330]: Let's inline this

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/ImportDeclarationNodeContext.java [hunk 2]
```diff
@@ -318,30 +319,59 @@ private ArrayList<LSCompletionItem> moduleNameContextCompletions(BallerinaComple
...
+            });
+            List<String> filteredPackages = getFilteredPackages(packageList, prefix, context);
+            for (String filteredPackage : filteredPackages) {
+                LSCompletionItem completionItem = getImportCompletion(context, filteredPackage, filteredPackage);
                 completionItem.getCompletionItem().setAdditionalTextEdits(additionalEdits);
                 completionItems.add(completionItem);
             }
```
Comment (malinthar) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/ImportDeclarationNodeContext.java:340]: Let's add a return here and get rid of the else block

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/CompletionSearchProvider.java [hunk 3]
```diff
@@ -85,4 +85,13 @@ public void indexModule(ModuleID moduleName, List<String> stringList, List<Strin
     public boolean checkModuleIndexed(ModuleID moduleName) {
         return indexedModules.contains(moduleName);
     }
+
+    /**
+     * Add central packages to the trie.
+     *
+     * @param packages    list of packages.
+     */
+    public void addCentralPackagesToTrie(List<String> packages) {
```
[the remark this answers is not in the document]
Comment (mindula) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/CompletionSearchProvider.java]: done

**Review by malinthar — CHANGES_REQUESTED** [review]

File: language-server/modules/langserver-core/src/test/resources/completion/import_decl/source/import_decl_with_ballerinax_org3.bal [hunk 4]
```diff
@@ -0,0 +1 @@
+import ballerinax/azure.
```
[the remark this answers is not in the document]
Comment (mindula) [reply language-server/modules/langserver-core/src/test/resources/completion/import_decl/source/import_decl_with_ballerinax_org3.bal:1]: +1

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/ImportDeclarationNodeContext.java (hunk 1 above)
Comment (KavinduZoysa) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/ImportDeclarationNodeContext.java:333]: It seems this is not fixed.

**Review by malinthar — APPROVED** [review]
LGTM!

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/ImportDeclarationNodeContext.java (hunk 1 above)
Comment (mindula) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/ImportDeclarationNodeContext.java]: Done.

**Review by malinthar — APPROVED** [review]

**Review by KavinduZoysa — APPROVED** [review]

**Pull request merged**